### PR TITLE
Revert "Revert "Short circuit optimization for functions executed over Nullable arguments"" 

### PR DESCRIPTION
**Pull request by taiyang-li**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Revert "Revert "Short circuit optimization for functions executed over Nullable arguments"" . Add fast path for functions with all argument types is numeric. Fix performance issues in  https://github.com/ClickHouse/ClickHouse/pull/72258

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing)

All builds in Builds_1 and Builds_2 stages are always mandatory
and will run independently of the checks below:

- [x]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64
- [ ]  Exclude: All with release
- [ ]  Exclude: All with debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 2ba5a71841f645ed0a39b9725eab716927f76828 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73820/2ba5a71841f645ed0a39b9725eab716927f76828/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/70ed5c7a995d316d2fbc85f8c5fa6875ac9375f2/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by taiyang-li** [comment]
@Avogar @Algunenano  I had added a fast path for functions with all argument type are numeric. And the performance tests  show no obvious degradations.

**Comment by taiyang-li** [comment]
slowdown: [1](https://s3.amazonaws.com/clickhouse-test-reports/73820/b1ae61a03366702fcea03ca6a2f3102e7db28790/performance_comparison__aarch64__[1_4]/report.html)
speedup:

**Comment by cwurm** [comment]
This change seems to cause these test failures: https://github.com/ClickHouse/ClickHouse/issues/72265

I tested the build right before this PR (from https://github.com/ClickHouse/ClickHouse/pull/73813) and the build from this PR. Both reproduction examples work with one but not the other.

**Comment by taiyang-li** [comment]
@cwurm I'll look into it next week.

**Comment by Algunenano** [comment]
I still don't understand the motivation behind this PR. We are adding hundreds of lines of complex code but we don't see any performance improvement. The mentioned [speedup](https://s3.amazonaws.com/clickhouse-test-reports/73820/b1ae61a03366702fcea03ca6a2f3102e7db28790/performance_comparison__aarch64__%5B4_4%5D/report.html) looks unrelated as they are not even using Nullable arguments. Are you sure the change is actually improving anything? 

Could you please create a performance test that verifies this change and shows a scenario where it's beneficial, please? Otherwise it's hard to justify keeping the extra complexity

File: src/Core/Settings.cpp [hunk 1]
```diff
@@ -5615,6 +5615,13 @@ Only available in ClickHouse Cloud. Number of background threads for speculative
 Use async and potentially multithreaded execution of materialized view query, can speedup views processing during INSERT, but also consume more memory.)", 0) \
     DECLARE(Int64, ignore_cold_parts_seconds, 0, R"(
 Only available in ClickHouse Cloud. Exclude new data parts from SELECT queries until they're either pre-warmed (see cache_populated_by_fetch) or this many seconds old. Only for Replicated-/SharedMergeTree.
+)", 0) \
+    DECLARE(Bool, short_circuit_function_evaluation_for_nulls, true, R"(
```
Comment (rschu1ze) [inline src/Core/Settings.cpp:5619]: As a developer, I find the description quite ambiguous. A normal user would not understand it at all. Kindly re-phrase and please add an example.

File: src/Functions/IFunction.cpp [hunk 2]
```diff
@@ -188,17 +202,101 @@ ColumnPtr IExecutableFunction::defaultImplementationForNulls(
...
 
+    if (null_presence.has_null_constant)
+    {
+        /// If any of the input arguments is null literal, the result is null constant.
         return result_type->createColumnConstWithDefaultValue(input_rows_count);
     }
 
     if (null_presence.has_nullable)
     {
+        /// Usually happens during analyzing. We should return non-const column to avoid wrong constant folding.
```
Comment (rschu1ze) [inline src/Functions/IFunction.cpp:215]: Uninitiated developers will find below code quite difficult. It has comments about the "how" but not the "why". Could you please add more context and motivation from the PR message of https://github.com/ClickHouse/ClickHouse/pull/60129 ?

**Comment by rschu1ze** [comment]
> I'll add it in another pr later.

Quick question: In which PR was the performance tests added?

**Comment by taiyang-li** [comment]
@rschu1ze I'm going to add performance tests and comments in https://github.com/ClickHouse/ClickHouse/pull/73780 this week.

**Pull request merged**